src/v/cloud_storage/remote_segment.cc

```diff
@@ -976,12 +976,16 @@ ss::future<> remote_segment::hydrate(storage::opt_abort_source_t as) {
 
     // A no-op abort source is created on heap so that `do_hydrate` can call
     // `with_timeout_abortable` if we do not have an input abort source.
-    return ss::do_with(ss::abort_source{}, [this, as](ss::abort_source& noop) {
-        return do_hydrate(
-          as.value_or(noop),
-          ss::lowres_clock::now()
-            + config::shard_local_cfg().cloud_storage_hydration_timeout_ms());
-    });
+    return ss::do_with(
+             ss::abort_source{},
+             [this, as](ss::abort_source& noop) {
+                 return do_hydrate(
+                   as.value_or(noop),
+                   ss::lowres_clock::now()
+                     + config::shard_local_cfg()
+                         .cloud_storage_hydration_timeout_ms());
+             })
+      .finally([holder = std::move(g)] {});
```

Comment:
Does this ensure that the gate will be held until the returned future is resolved?

Reply:
Yeah, the units are copied and held onto until the copy is destructed (implicitly, in the body of the `finally()`